directdemod/georeferencer.py

```diff
@@ -0,0 +1,215 @@
+                                    tps=True,
+                                    resampleAlg=GRA_NearestNeighbour)
+
+        gdal.Warp(destNameOrDestDS=output_file,
+                    srcDSOrSrcDSTab=constants.TEMP_TIFF_FILE,
+                    options=options)
+
+        os.remove(constants.TEMP_TIFF_FILE)
+
+        if desc:
+            self.create_desc(descriptor, output_file)
+
+    def georef_os(self, descriptor, output_file, desc=False):
+
+        '''Main georeferencing routine
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+            desc (:obj:`bool`): descriptor flag
+        '''
+
+        file_name = descriptor["image_name"]
+        image     = mimg.imread(file_name)
+
+        gcps = self.compute_gcps(descriptor, image)
+
+        translate_flags = "-of GTiff -a_srs EPSG:4326"
+        warp_flags = "-r near -tps -s_srs EPSG:4326 -t_srs EPSG:4326"
+
+        translate_query = 'gdal_translate ' + translate_flags + ' ' + self.to_string_gcps(gcps) + ' "' + file_name + '" ' + ' "' + constants.TEMP_TIFF_FILE + '"'
+        warp_query = 'gdalwarp ' + warp_flags + ' "' + constants.TEMP_TIFF_FILE + '" ' + ' "' + output_file + '"'
+
+        os.system(translate_query)
+        os.system(warp_query)
+
+        os.remove(constants.TEMP_TIFF_FILE)
+
+        if desc:
+            self.create_desc(descriptor, output_file)
+
+    def to_string_gcps(self, gcps):
+
+        '''Create string representation of gcp points
+
+        Args:
+            gcps (:obj:`list`): list of gcp points
+        '''
+
+        return " ".join([("-gcp " + str(gcp.GCPPixel) + " " + str(gcp.GCPLine) + " " + str(gcp.GCPX) + " " + str(gcp.GCPY)) for gcp in gcps])
+
+
+    def create_desc(self, descriptor, output_file):
+
+        '''Create descriptor file
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+        '''
+
+        desc = {
+            "image_name": output_file,
+            "sat_type": descriptor["sat_type"],
+            "date_time": descriptor["date_time"],
+            "center": descriptor["center"],
+            "direction": 0
+        }
+
+        name, extension = os.path.splitext(output_file)
+        desc_name = name + "_desc.json"
+        JsonParser.save(desc, desc_name)
+
+    def compute_gcps(self, descriptor, image):
+
+        '''Compute set of GCPs
+
+        Args:
+            h (:obj:`dict`): descriptor dictionary
+            w (:obj:`np.ndarray`): image as np.ndarray
+        '''
+
+        height = image.shape[0]
+        width  = image.shape[1]
+        center_w = width/2
+        center_h = height/2
+
+        gcps = []
+        dtime = dparser.parse(descriptor["date_time"])
+        orbiter = Orbital(descriptor["sat_type"], tle_file=self.tle_file)
+        min_delta = 500
+        middle_dist = 3.22 * 455 / 2. * 1000
```

Comment:
could you please explain what these middle_dist and far_dist are? how did you arrive at these numbers?

Reply:
middle_dist is quarter length of image. So image an image with height 1000 and width 910, so the quarter is 455 / 2 pixels, and I multiply it by 3.22 * 1000, 3.22 km in pixel and convert it to meters. far_dist is the half of the image.
If you stay in the middle of the image, far_dist is the distance to the left and right sides, middle_dist is half of that distance.